client/mysqlbinlog.cc

```diff
@@ -744,11 +753,64 @@ static void convert_path_to_forward_slashes(char *fname)
 */
 static bool shall_skip_database(const char *log_dbname)
 {
-  return one_database &&
-         (log_dbname != NULL) &&
-         strcmp(log_dbname, database);
+  if (log_dbname == NULL)
+  {
+    return false;
+  }
+
+  if (one_database)
+    return strcmp(log_dbname, database);
+  else if (multi_databases)
+  {
+    return filter_databases.count(log_dbname) == 0;
+  }
+  else
+    return false;
 }
 
+/**
+  Indicates whether the given database should be filtered out,
+  according to the --flashback-databases=X,X,X option.
+
+  @param log_dbname Name of database.
+
+  @return nonzero if the database with the given name should be
+  filtered out, 0 otherwise.
+*/
+static bool flashback_shall_skip_database(const char *log_dbname)
+{
+    if (log_dbname == NULL){
+        return false;
+    }
+
+    if (flashback_multi_databases){
+        return flashback_filter_databases.count(log_dbname) == 0;
+    }
+    else
+        return false;
+}
+
+/**
+  Indicates whether the given table should be filtered out,
+  according to the --flashback-tables=X,X,X option.
+
+  @param log_tblname Name of table.
+
+  @return nonzero if the table with the given name should be
+  filtered out, 0 otherwise.
+*/
+static bool flashback_shall_skip_table(const char *log_tblname)
```

Comment:
I may have misunderstood how these options work, If that's the case, let me know and apologies in advance. I have some doubts regarding the extra options you are trying to add:

I don't understand why we would need to add a separate option for `--databases`, instead of using the existing `--database` option modified to accept a list.
Also, is there a reason why we would need the `flashback-databases` option as well, it seems to me like this one can also be piggybacked on the existing `--databases` option.
For the `--flashback-tables` option there is no existing candidate, but if it's not too much extra work, we should rename it to `--tables` and make it work even when the `--flashback` option is not present.

Reply:
Thanks, robert. Let me introduce the background of  `--databases`  `--flashback-databases`  `--flashback-tables` options.
1.  "Using the existing --database option modified to accept a list" is ok.
However, according to the experience of most people around me `--database` generally means receiving a single database, `--databases` means receiving a list of databases.
2. 
- In our production environment, it is a common requirement for flashback to process multiple databases and multiple tables at the same time. On the contrary, it is very rare for flashback to handle single database and single table.

- The `--tables` option cannot be added, because when `binlog_format=statement`, the binlog file records sql operations through Query_event, and we cannot get the target table from Query_event.
Although `binlog_format=row`, DDL (create/alter/drop) is still saved in Query_event , we cannot  get target table from Query_event.
- `--flashback` can only be used when binlog_format=row, and only processes DML (insert/update/delete), ignoring DDL (create/alter/drop) ) Query_event, so adding `--flashback-tables` option is ok.
- `--flashback-databases` and `--databases` have the same effect,  in my opinion `--flashback-databases` + `--flashback-tables` seems to be more friendly. 
But according to your suggestions, I can remove `--flashback-databases` related code.
- It should be noted that the `--table` option is only valid for DML (insert/update/delete) when `binlog_format=row`. If in a production environment, `binlog_format=row` + DDL or `binlog_format=statement`, mysqlbinlog --table will cause a big problem.

Reply:
> I may have misunderstood how these options work, If that's the case, let me know and apologies in advance. I have some doubts regarding the extra options you are trying to add:
> 
> I don't understand why we would need to add a separate option for `--databases`, instead of using the existing `--database` option modified to accept a list.
> Also, is there a reason why we would need the `flashback-databases` option as well, it seems to me like this one can also be piggybacked on the existing `--databases` option.
> For the `--flashback-tables` option there is no existing candidate, but if it's not too much extra work, we should rename it to `--tables` and make it work even when the `--flashback` option is not present.

Thanks, robert . Let me introduce the background of  `--databases`  `--flashback-databases`  `--flashback-tables` options.
1.  "Using the existing --database option modified to accept a list" is ok.
However, according to the experience of most people around me `--database` generally means receiving a single database, `--databases` means receiving a list of databases.
2. 
- In our production environment, it is a common requirement for flashback to process multiple databases and multiple tables at the same time. On the contrary, it is very rare for flashback to handle single database and single table.

- The `--tables` option cannot be added, because when `binlog_format=statement`, the binlog file records sql operations through Query_event, and we cannot get the target table from Query_event.
Although `binlog_format=row`, DDL (create/alter/drop) is still saved in Query_event , we cannot  get target table from Query_event.
- `--flashback` can only be used when binlog_format=row, and only processes DML (insert/update/delete), ignoring DDL (create/alter/drop) ) Query_event, so adding `--flashback-tables` option is ok.
- `--flashback-databases` and `--databases` have the same effect,  in my opinion `--flashback-databases` + `--flashback-tables` seems to be more friendly. 
But according to your suggestions, I can remove `--flashback-databases` related code.

Reply:
I just replied with the same suggestion as @robertbindar has.
I would agree the plural form `--databases, --tables` would be nicer, still the singular versions
are not that uncommon (in practice). Like I've said in my comments the plural forms still must work with the single item list.

To
> The --tables option cannot be added
while you're right about `binlog_format=statement` this must be *not* an obstacle. The statement format just won't work out (as of current) with `--table=t1,t2`, like would not work out
the suggested `--flashback-tables=t1,t2`.
So to me it's fine to have the existing option name in singular like `--table=t1,t2` and it would be limited to `ROW` format at the moment.
So that would be a *general* feature while the backward compatibility is in place.